### ISSUE TITLE
compute loudness error

### ISSUE THREAD

**Issue by james20141606** [magenta/ddsp#56]
Hey, I am trying to compute loudness from an only 4 seconds audio. And it turns out that the loudness calculation just failed.
![image](https://user-images.githubusercontent.com/14315324/77025763-5615b580-6968-11ea-9403-bc83e24a6290.png)
Is it because the audio is too short? Could I just use spectral_ops.compute_loudness on multiple 4 seconds long audio samples separately?

**Comment by james20141606** [comment]
I am also not sure about the f0 estimation. I know one way is to use CREPE to compute f0, and in ae_abs, we can set to train a ResnetF0Encoder for f0 encoding.
I have trained an autoencoder and I would like to use its ResnetF0Encoder module to estimate f0, but I found the f0 estimated looks different from f0 estiimated by CREPE, should I use CREPE's result as my first choice?

![image](https://user-images.githubusercontent.com/14315324/77029638-9da13f00-6972-11ea-86c4-51e3577dd0bd.png)

**Comment by jesseengel** [comment]
Perhaps you need to normalize your input audio? 1e-11 and 1e-10 seem very small and should be imperceivable.

In terms of the f0 encoder, the CREPE model is the only reliable one at the moment. As we said in the paper, the f0 autoencoder can learn to reproduce the audio, but estimating the correct f0 in the process is still ongoing work.

**Comment by james20141606** [comment]
Thx for the reply!
For the audio value, I think it is small because I use audio /= 2**(8 * audio_segment.sample_width) in _load_audio_as_array function. So do you mean it might be better not use this part?
I also checked the codes and it seems that the pipeline estimate the f0 and loudness of a full length audio and segment it into 4 seconds pieces, is it right? 
For f0 encoder, I mean if I use ae_abs.gin to pretrain an autoencoder, and I think it jointly trains a f0 encoder. Then if I want to apply this pretrained f0 encoder to extract f0 instead CREPE model, is it ok (better)?

**Comment by jesseengel** [comment]
> For the audio value, I think it is small because I use audio /= 2**(8 * audio_segment.sample_width) in _load_audio_as_array function. So do you mean it might be better not use this part?

This is to convert the int values to float, which is usually required for wav files, but maybe not for others? What files types are you using for your dataset?

You can check the format of the loaded data with `audio_segment.array_type`. You could add a line to skip the integer-> float if array type is 'f' or 'd' according to this table:

```
    Type code   C Type             Minimum size in bytes
    'b'         signed integer     1
    'B'         unsigned integer   1
    'u'         Unicode character  2 (see note)
    'h'         signed integer     2
    'H'         unsigned integer   2
    'i'         signed integer     2
    'I'         unsigned integer   2
    'l'         signed integer     4
    'L'         unsigned integer   4
    'q'         signed integer     8 (see note)
    'Q'         unsigned integer   8 (see note)
    'f'         floating point     4
    'd'         floating point     8
```

See if that works for you, and if so, please include a patch :).

Like I said above, getting real f0 values out of the f0 encoder is ongoing research, so pretrained CREPE is better for now.

**Comment by james20141606** [comment]
I loaded some audio stored in .mat file so I don't know the original file format, it is already float so I think I do not need to convert it to float.

I just use the `add_loudness` and `add_f0_estimate` again and from my understanding, it is for audio segment (4 seconds long pieces) instead of the whole length audio, right? but in `prepare_tfrecord` function it seems the `add_loudness` and `add_f0_estimate` process the whole length audio at first and then split it into small pieces? I am not sure if the processing order will influence the result.

I also noticed that the silence period length of the audio can influence the audio extraction if the silent period is longer, the extracted feature seems to loss more details, should we ignore it or try to fix it?

**Comment by jesseengel** [comment]
Ah, yes, that script is for loading from audio files, so that's your problem, no conversion necessary.

The processing order shouldn't influence the result greatly (we do use viterbi decoding but there aren't super long term dependencies). You can also choose what size you would like to split and hop as top_level flags.

The extracted features should be insensitive to padding with silence, as long as the audio is sufficiently loud.

**Comment by james20141606** [comment]
Thanks for the explanation! For the last problem, I test on two audios like this:

![image](https://user-images.githubusercontent.com/14315324/77120068-5072ab00-6a0e-11ea-8223-5d9f39f28759.png)
![image](https://user-images.githubusercontent.com/14315324/77120076-54063200-6a0e-11ea-9c1a-3dd734329e9c.png)
Seems that if the silent period is longer, the extracted features are not very precise?

**Comment by jesseengel** [comment]
It's hard for me to tell without hearing the audio / reconstructions, but I
think it may be doing the right thing. It seems like you're using two
different pieces of audio. The loudness contours are similar. The F0 varies
a lot for the longer sample, in the silent parts, but keep in mind that the
f0_confidence is also low so those are too be ignored. If there is a
problem, it seems to be more with CREPE than anything else. Also, you want
to make sure you're using audio with 16kHz sampling rate for CREPE.

**Comment by james20141606** [comment]
thank you for your help!
